src/NetworkPaymentAdapter.sol

```diff
@@ -0,0 +1,134 @@
+        wards[usr] = 1;
+
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "Sequencer/not-authorized");
+        _;
+    }
+
+    // --- Data ---
+    VestLike public immutable vest;
+    uint256 public immutable vestId;
+    INetworkTreasury public immutable treasury;
+    DaiJoinLike public immutable daiJoin;
+    DaiLike public immutable dai;
+    address public immutable vow;
+
+    // --- Parameters ---
+    uint256 public bufferMax;
+    uint256 public minimumPayment;
+
+    // --- Tracking ---
+    uint256 public totalSent;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, uint256 data);
+    event TopUp(uint256 bufferSize, uint256 daiBalance, uint256 daiSent, uint256 totalSent);
+
+    // --- Errors ---
+    error InvalidFileParam(bytes32 what);
+    error BufferFull(uint256 bufferSize, uint256 bufferMax);
+    error PendingDaiTooSmall(uint256 pendingDai, uint256 minimumPayment);
+
+    constructor(address _vest, uint256 _vestId, address _treasury, address _daiJoin, address _vow) {
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+
+        vest = VestLike(_vest);
+        vestId = _vestId;
+        treasury = INetworkTreasury(_treasury);
+        daiJoin = DaiJoinLike(_daiJoin);
+        dai = DaiLike(daiJoin.dai());
+        vow = _vow;
+    }
+
+    // --- Administration ---
+    function file(bytes32 what, uint256 data) external auth {
+        if (what == "bufferMax") {
+            bufferMax = data;
+        } else if (what == "minimumPayment") {
+            minimumPayment = data;
+        } else revert InvalidFileParam(what);
+
+        emit File(what, data);
+    }
+
+    // --- Pay the keeper treasury ---
+    function topUp() external returns (uint256 daiSent) {
+        uint256 bufferSize = treasury.getBufferSize();
+        uint256 pendingDai = vest.unpaid(vestId);
+        uint256 _bufferMax = bufferMax;
+        uint256 _minimumPayment = minimumPayment;
+
+        if (bufferSize + _minimumPayment >= _bufferMax) revert BufferFull(bufferSize, _bufferMax);
+        else if (pendingDai >= _minimumPayment) {
+            vest.vest(vestId);
+            
+            // Send DAI up to the maximum and the rest should go back into the surplus buffer
+            // Use the balance in case someone sends DAI directly to this contract (can be used in emergency)
+            uint256 daiBalance = dai.balanceOf(address(this));
+            if (daiBalance + bufferSize > _bufferMax) {
+                daiSent = _bufferMax - bufferSize;
+
+                // Send the rest back to the surplus buffer
+                daiJoin.join(vow, daiBalance - daiSent);
+            } else {
+                daiSent = daiBalance;
+            }
+            dai.transfer(address(treasury), daiSent);
```

Comment:
Our system is quite flexible so we can make it work in whatever way, but the easiest and most gas efficient would be the following: 

We have a central `TaskTreasury` [contract](https://github.com/gelatodigital/ops/blob/ea4f0dcb023861bce9ebf0840460b674cae04874/contracts/TaskTreasury/TaskTreasuryUpgradable.sol#L136) which collects the funds of all users. This contract has a function that needs to be called specifying MakerDAO's Task Creator address as the beneficiary in order to top up the right balance.

instead of `dai.transfer(address(treasury), daiSent);`

the best way for us would be smth like we currently do: 

```
dai.approve(address(treasury), daiSent);
treasury.depositFunds(makerDAO, address(dai), daiSent);
```

The second part which is Gelato specific could also be done with a low level call and pre-defined bytes (except the last value) in order to make it generic for other protocols (i.e. via a whitelisted function that gets called per protocol). 

If that won't work, then we would just create a wrapper contract like we do now which calls the `topUp` function and then deposits it into Maker's Gelato Balance from there.

Reply:
I think we can just do the wrapper contract which calls `topUp()`, so the `NetworkPaymentAdapter` doesn't need to be aware of the interface of the other contract which could vary between networks in any number of ways.